src/main/java/co/cask/gcp/datastore/sink/DatastoreSink.java

```
@@ -0,0 +1,343 @@
+import co.cask.gcp.datastore.sink.util.SinkKeyType;
+import co.cask.gcp.datastore.util.DatastorePropertyUtil;
+import co.cask.hydrator.common.LineageRecorder;
+import com.google.cloud.Timestamp;
+import com.google.cloud.datastore.Blob;
+import com.google.cloud.datastore.BlobValue;
+import com.google.cloud.datastore.BooleanValue;
+import com.google.cloud.datastore.DoubleValue;
+import com.google.cloud.datastore.EntityValue;
+import com.google.cloud.datastore.FullEntity;
+import com.google.cloud.datastore.IncompleteKey;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.LongValue;
+import com.google.cloud.datastore.NullValue;
+import com.google.cloud.datastore.PathElement;
+import com.google.cloud.datastore.StringValue;
+import com.google.cloud.datastore.TimestampValue;
+import com.google.cloud.datastore.Value;
+import com.google.common.base.Strings;
+import org.apache.hadoop.io.NullWritable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link BatchSink} that writes data to Cloud Datastore.
+ * This {@link DatastoreSink} takes a {@link StructuredRecord} in, converts it to Entity, and writes it to the
+ * Cloud Datastore kind.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name(DatastoreSink.PLUGIN_NAME)
+@Description("CDAP Google Cloud Datastore Batch Sink takes the structured record from the input source and writes "
+  + "the data into Google Cloud Datastore.")
+public class DatastoreSink extends BatchSink<StructuredRecord, NullWritable, FullEntity<?>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatastoreSink.class);
+
+  public static final String PLUGIN_NAME = "Datastore";
+
+  private final DatastoreSinkConfig config;
+
+  public DatastoreSink(DatastoreSinkConfig config) {
+    this.config = config;
+  }
+
```

configurePipeline() should be implemented, where it validates properties and the input schema. This will cause invalid pipelines to fail early when deployed, rather than fail at runtime.

---

Done.